ironic/resource_ironic_node_v1.go

```diff
@@ -585,3 +614,131 @@ func changePowerState(client *gophercloud.ServiceClient, d *schema.ResourceData,
 
 	return nil
 }
+
+// setRAIDConfig calls ironic's API to send request to change a Node's RAID config.
+func setRAIDConfig(client *gophercloud.ServiceClient, d *schema.ResourceData) (err error) {
+	var logicalDisks []nodes.LogicalDisk
+	var raid *metal3v1alpha1.RAIDConfig
+
+	raidConfig := d.Get("raid_config").(string)
+	if raidConfig == "" {
+		return nil
+	}
+
+	err = json.Unmarshal([]byte(raidConfig), &raid)
+	if err != nil {
+		return
+	}
+
+	err = checkRAIDConfigure(d.Get("raid_interface").(string), raid)
+	if err != nil {
+		return
+	}
+
+	// Build target for RAID configuration steps
+	logicalDisks, err = ironic.BuildTargetRAIDCfg(raid)
+	if len(logicalDisks) == 0 || err != nil {
+		return
+	}
+
+	// Set root volume
+	if len(d.Get("root_device").(map[string]interface{})) == 0 {
+		logicalDisks[0].IsRootVolume = new(bool)
+		*logicalDisks[0].IsRootVolume = true
+	} else {
+		log.Printf("rootDeviceHints is used, the first volume of raid will not be set to root")
+	}
+
+	// Set target for RAID configuration steps
+	return nodes.SetRAIDConfig(
+		client,
+		d.Id(),
+		nodes.RAIDConfigOpts{LogicalDisks: logicalDisks},
+	).ExtractErr()
+}
+
+// buildBIOSSettings builds bios settings for BIOS clean steps
+func buildBIOSSettings(d *schema.ResourceData, firmwareConfig *bmc.FirmwareConfig) (settings []map[string]string, err error) {
+	driver := d.Get("driver").(string)
+	driverInfo := d.Get("driver_info").(map[string]interface{})
+	driverAddress := strings.Join([]string{driver, "address"}, "_")
+
+	address := strings.Join([]string{driver, driverInfo[driverAddress].(string)}, "://")
+	acc, err := bmc.NewAccessDetails(address, false)
+	if err != nil {
+		return nil, err
+	}
+
+	settings, err = acc.BuildBIOSSettings(firmwareConfig)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+// buildManualCleaningSteps builds the clean steps for RAID and BIOS configuration
+func buildManualCleaningSteps(d *schema.ResourceData) (cleanSteps []nodes.CleanStep, err error) {
+	var targetRaid *metal3v1alpha1.RAIDConfig
+	var firmware *bmc.FirmwareConfig
+
+	raidInterface := d.Get("raid_interface").(string)
+
+	raidConfig := d.Get("raid_config").(string)
+	if raidConfig != "" {
+		if err = json.Unmarshal([]byte(raidConfig), &targetRaid); err != nil {
+			return nil, err
+		}
+
+		// Build raid clean steps
+		raidCleanSteps, err := ironic.BuildRAIDCleanSteps(raidInterface, targetRaid, nil)
+		if err != nil {
+			return nil, err
+		}
+		cleanSteps = append(cleanSteps, raidCleanSteps...)
+	}
+
+	biosSetings := d.Get("bios_settings").(string)
+	if biosSetings != "" {
+		if err = json.Unmarshal([]byte(biosSetings), &firmware); err != nil {
+			return nil, err
+		}
+
+		settings, err := buildBIOSSettings(d, firmware)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(settings) != 0 {
+			cleanSteps = append(
+				cleanSteps,
+				nodes.CleanStep{
+					Interface: "bios",
+					Step:      "apply_configuration",
+					Args: map[string]interface{}{
```

Comment:
Did you run gofmt prior?, I'd expect space between curly brackets?

Reply:
Yes, I have run. This is not a fmt issue.

Reply:
I think this is actually correct per Go conventions, although I also find it annoying :)